README.md
aes-crypto-api-iface
====================

This directory contains sources of driver that we use to test FPGA firware
providing AES decryption hardware accelerator.

It uses standard kernel Crypto-API interface. See:
  https://kernel.readthedocs.io/en/sphinx-samples/crypto-API.html


FILES
-----

 * `aes-crypto-api-iface.c` -- the driver source code
 * `socfpga_cyclone5_etln.dts` -- sample DTS file containing decription of
   platform device handled by the driver

BUILD
-----

Building the module is straingtforward:

```shell
make KDIR=path/to/prebuilt/kernel
```

KDIR must contain a path to kernel sources.

In addition, one may want to pass `CROSS_COMPILE` parameter to `make` to change
cross compilation toolchain prefix. The default `CROSS_COMPILE` value is
`arm-linux-gnueabihf-`. Make sure you build the module with the same toolchain
that was used to build a kernel.

INSTALL
-------

After running the `make` described in the previous section you will obtain
`aes-crypto-api-iface.ko` in the current directory. You may want to place it in
`/usr/lib/modules/<kernel-version>/` on target machine and run `depmod` making
the module load automatically once the kernel finds appropriate platform
device. Or instead you can simply copy it to whatever directory on the target
machine and `insmod` by hand.

The module provides a platform device driver that handles devices having
`compatible` property equal to `stcmtk,aes`. For example, such device can be
described in dts by the following lines:

```dts
my-aes-decryptor {
	compatible = "stcmtk,aes";
	interrupts = <0x0 0x1 0x2   /* decryption irq */
		      0x0 0x1 0x3>; /* encryption irq */
};
```

Currently all the device properties (register addresses, block size) are
hardcoded in the driver, so no properties except `compatible` and `interrupts`
can be specified in DTS. `interrputs` property must describe two interrupts
that is issued by decryption and encryption accelerators' DMA controllers in
FPGA respectively.

STC Metortek's SoC-based devices are shipped with custom fpga manager driver
(`etn-fpga-mgr.ko`) installed while the upstream Altera's fpga manager (called
`altera.ko` or `socfpga.ko`) device's node is disabled in dts. The Metrotek's
fpga manager driver requires the FPGA firmware to have special capability
called 'features'.

At the moment of writing `aes-crypto-api-iface` driver the firmware providing
hardware AES decryption accelerator device did not have 'features'. In order to
run the firmware properly you have to disable the device handled by Metrotek's
FPGA manager and enable Altera FPGA manager's device.

The task of enabling proper FPGA manager and disabling the other as well as
describing device tree node for the device handled by `aes-crypto-api-iface` are
solved in `socfpga_cyclone5_etln.dts` DTS for ETLN that can be found in this
directory. Make sure you boot with proper DTS.

USAGE
-----

Once you have compiled and installed the `aes-crypto-api-iface`, installed the
proper FPGA firmware and booted the machine with proper DTS you may load the
`aes-crypto-api-iface` module and start having fun.

Once loaded the module registers an in-kernel cryptographic cipher that can be
used by various consumers mostly in kernel. Userspace can put our hardware
cipher to use through interfaces like `AF_ALG` socket family [2] or
`cryptodev.ko` out-of-tree module [1].

Minimal usage example of our module is presented below. It uses openssl in
conjunction with `cryptodev.ko` to decrypt using hardware acceleration and
compare with reference result. The openssl must be compiled with cryptodev
engine support:
  http://forum.doozan.com/read.php?2,18152

```shell
# Ensure that our Crypto device is registered:
cat /proc/crypto | grep -A 10 fpga

# Make tmpfs
mkdir tmpfs
mount -t tmpfs -o size=300m,nr_inodes=100,mode=0700 tmpfs tmpfs/
cf tmpfs

# Genarate random ciphertext, key and initialization vector
dd bs=16000 count=1000 if=/dev/urandom of=ciphertext
dd bs=1 count=16 if=/dev/urandom of=key
dd bs=1 count=16 if=/dev/urandom of=iv

# Decrypt ciphertext without FPGA acceleration (make sure that
# `aes-crypto-api-iface` and `cryptodev` modules are not loaded)
time openssl aes-128-cbc -d -in ciphertext -out nofpga_plaintext -K $(xxd -p key) -nopad -iv $(xxd -p iv)

# Load modules
insmod <your-path>/cryptodev.ko
insmod <your-path>/aes-crypto-api-iface.ko

# Decrypt ciphertext with FPGA acceleration
time openssl aes-128-cbc -d -in ciphertext -out fpga_plaintext -K $(xxd -p key) -nopad -iv $(xxd -p iv)

# Compare results
diff fpga_plaintext nofpga_plaintext
```

The example above insmod's and rmmod's modules to switch between hardware and
software encryption. We hope that there is a more convenient way to achieve
this, but it is not known to us.

SEE ALSO
--------

 1. https://github.com/cryptodev-linux/cryptodev-linux
 2. https://kernel.readthedocs.io/en/sphinx-samples/crypto-API.html


ChangeLog
aes-crypto-api-iface (0.0.1-1) unstable; urgency=low
  2017/01/27
    ! aes-crypto-api-iface.c:
        Fix -- we use a separate iv for each operation mode (encrypt & decrypt).
      [dg]

  2017/01/27
    + aes-crypto-api-iface.c:
        1) Add encrypt operation support.
           It is use separate control registers set for DMA & Crypto engine
           (include registers for key & iv).
           Now we write same key & iv in both (encrypt & decrypt) registers.

        2) Delete reading IRQ status in ISR.
           Now it is not necessary - FPGA has become more smart.

        3) Comment out obtaining IRQ numbers from .dtb
           Hardcoded IRQ to 72 for decrypt engine & 73 for encrypt engine.
           It is necessary to work with debug FPGA-firmware.
      [dg]

  2017/01/26
    ! aes-crypto-api-iface.c:
        Bugfix -- add copying decrypted plain text from temporary
        buffer to destination scatterlist buffers.
      [dg]

    ! aes-crypto-api-iface.c:
        Fix bug in temporary buffer offset calculation.
      [dg]

  2017/01/18
    * aes-crypto-api-iface.c:
        Fix NULL pointer dereference bug caused by ISR being registered before
        resources allocation.
      [io]

  2017/01/18
    * aes-crypto-api-iface.c:
        General code cleanup fixing indentation and braces placement.
      [io]

  2017/01/18
    * socfpga_cyclone5_etln.dts:
        Describe interrupt used by AES-accelerator's DMA controller.
      [io]

  2017/01/17
    + netdma.h,
    * aes-crypto-api-iface.c:
        Rewrite driver to make use of new features of FPGA firmware that
        include scatter-gather DMA and CBC mode of operation.
      [dg, io]


 -- Ivan Oleynikov <io@metrotek.spb.ru>  Wed, 18 Jan 2017 17:11:01 +0300

aes-crypto-api-iface (0.0.1-0) unstable; urgency=low
  2016/12/16
    + README.md:
        Few minor changes of README
      [io]

  2016/12/15
    + aes-crypto-api-iface.c:
        Initial version of AES-accelerator driver that supports
        in-kernel Crypto API interface.

        It supports only decryption now.

        It does not provide multi-block cipher and provides only
        single block cipher (see CRYPTO_ALG_TYPE_CIPHER in
        include/linux/crypto.h)

        See README.
      [dg, io]

    ! aes-accelerator.c:
        Delete old version.
      [dg, io]


 -- Denis Gabidullin <d.gabidullin@metrotek.spb.ru>  Thu, 15 Dec 2016 14:50:40 +0300


netdma.h
#ifndef NETDMA__H
#define NETDMA__H

#define NETDMA_CSR_SIZE 32

struct netdma_csr {
        u32 control;
        u32 status;
        u32 tx_status;
        u32 rx_report;
        u32 tx_desc_buf;
        u32 rx_desc_buf;
} __attribute__ ((packed, aligned(NETDMA_CSR_SIZE)));


struct netdma_rx_report{
  u16 actual_bytes_transferred;
  u8 sequence_number;   
  u8 flags;
};

struct netdma_tx_stat {
  u8 howmanydone;
  u8 sequence_number;
  u8 flags;
};
  

// Control
#define CTRL_RESET                                  BIT(0)

#define CTRL_RX_IRQ_ENABLE                          BIT(1) 
#define CTRL_TX_IRQ_ENABLE                          BIT(2) 

#define CTRL_CLEAR_RX_IRQ_STATUS                    BIT(3) 
#define CTRL_CLEAR_TX_IRQ_STATUS                    BIT(4) 


// Status
#define STAT_TX_IS_ANY_DONE                         BIT(0)

#define STAT_RX_REPORT_BUFFER_EMPTY                 BIT(1)

#define STAT_RX_IRQ_PENDING                         BIT(2)
#define STAT_TX_IRQ_PENDING                         BIT(3)

#define STAT_RX_DESC_BUFFER_FULL                    BIT(4) 
#define STAT_TX_DESC_BUFFER_FULL                    BIT(5)


// TX status
#define TXSTAT_DONE_CNT_OFFSET                      (0)
#define TXSTAT_DONE_CNT_MASK                        (0xFF)

#define TXSTAT_LAST_SEQ_NUM_OFFSET                  (8)
#define TXSTAT_LAST_SEQ_NUM_MASK                    (0xFF)


// RX report
#define RX_REPORT_ACTUAL_BYTES_OFFSET               (0)
#define RX_REPORT_ACTUAL_BYTES_MASK                 (0x3FFF)


// Descriptor 
#define DESC_BYTECOUNT_OFFSET                       (0) 
#define DESC_BYTECOUNT_MASK                         (0x3FFF) 

#define DESC_DISABLE_IRQ_OFFSET                     (14) 
#define DESC_DISABLE_IRQ_MASK                       (0x1) 

#define DESC_SEQ_NUM_OFFSET                         (16) 
#define DESC_SEQ_NUM_MASK                           (0xFF) 

#define TX_TIMESTAMP_OFFSET                         (24) 
#define TX_TIMESTAMP_MASK                           (0x1) 

#endif 


aes-accelerator.c
#include <linux/module.h>
#include <linux/kernel.h>
#include <linux/platform_device.h>
#include <linux/of.h>
#include <linux/printk.h>
#include <linux/device.h>

#include <linux/dma-mapping.h>
#include <linux/slab.h>

#include <linux/delay.h>

#include <linux/kdev_t.h>

#define FPGASLAVES (0xC0000000)

#define AES_BASE  FPGASLAVES
#define AES_SIZE  (7 * 4)

#define AES_BLOCK_SIZE 16
#define AES_KEY_SIZE   (32 * 4 / 8)

static struct class *encryptor_class;

struct aes_regs {
	u32 main_ctrl;
	u32 dma_read;
	u32 dma_write;
	u32 key[4];
	u32 block_counter;
} __attribute__((__packed__));

struct aes_priv {
	struct device *dev;
	struct aes_regs __iomem *regs;

	char *cipher_buf, *plain_buf;

	dma_addr_t cipher_dma, plain_dma;
};

static ssize_t key_store(struct device *dev, struct device_attribute *attr,
		const char *buf, size_t count)
{
	struct aes_priv *priv;
	const uint32_t *w_buf;

	priv = dev_get_drvdata(dev);

	dev_info(dev, "key = %s", buf);

	if (count != AES_KEY_SIZE) {
		dev_warn(dev, "Provided key of length %u when %u expected",
			(unsigned int) count, (unsigned int) AES_KEY_SIZE);
		return -EINVAL;
	}

	w_buf = (const uint32_t *)buf;

	iowrite32(w_buf[0], priv->regs->key);
	iowrite32(w_buf[1], priv->regs->key + 1);
	iowrite32(w_buf[2], priv->regs->key + 2);
	iowrite32(w_buf[3], priv->regs->key + 3);

	iowrite32(0x4, &priv->regs->main_ctrl);
	iowrite32(0x0, &priv->regs->main_ctrl);

	dev_info(dev, "key written successfully");

	return count;
}
DEVICE_ATTR_WO(key);

static ssize_t ciphertext_store(struct device *dev, struct device_attribute *attr,
		const char *buf, size_t count)
{
	struct aes_priv *priv;

	priv = dev_get_drvdata(dev);

	dev_info(dev, "ciphertext = %s", buf);

	if (count != AES_BLOCK_SIZE) {
		dev_warn(dev, "Got %u bytes of ciphertext when expected %u",
			(unsigned int) count, (unsigned int) AES_BLOCK_SIZE);
		return -EINVAL;
	}

	dma_sync_single_for_cpu(priv->dev, priv->cipher_dma, AES_BLOCK_SIZE, DMA_TO_DEVICE);

	memcpy(priv->cipher_buf, buf, AES_BLOCK_SIZE);

	dma_sync_single_for_device(priv->dev, priv->cipher_dma, AES_BLOCK_SIZE, DMA_TO_DEVICE);

	return count;
}
DEVICE_ATTR_WO(ciphertext);

static ssize_t plaintext_show(struct device *dev, struct device_attribute *attr, char *buf)
{
	struct aes_priv *priv;

	priv = dev_get_drvdata(dev);

	iowrite32(0x1, &priv->regs->main_ctrl);

	//msleep(1);

	dma_sync_single_for_cpu(priv->dev, priv->plain_dma, AES_BLOCK_SIZE, DMA_FROM_DEVICE);

	memcpy(buf, priv->plain_buf, AES_BLOCK_SIZE);

	dma_sync_single_for_device(priv->dev, priv->plain_dma, AES_BLOCK_SIZE, DMA_FROM_DEVICE);

	return AES_BLOCK_SIZE;
}
DEVICE_ATTR_RO(plaintext);

static const struct attribute *aes_attrs[] = {
	&dev_attr_key.attr,
	&dev_attr_ciphertext.attr,
	&dev_attr_plaintext.attr,
	NULL
};

static const struct attribute_group aes_attr_group = {
	.attrs = (struct attribute **) aes_attrs,
};

static int aes_probe(struct platform_device *pdev)
{
	int err;
	struct aes_priv *priv;
	struct device *dev;

	dev_info(&pdev->dev, "probing");

	priv = devm_kzalloc(&pdev->dev, sizeof(*priv), GFP_KERNEL);
	BUG_ON(!priv);

	dev = device_create(encryptor_class, &pdev->dev, MKDEV(0, 0), NULL,
			"aes");
	BUG_ON(IS_ERR(priv->dev));

	priv->dev = dev;
	dev_set_drvdata(dev, priv);
	platform_set_drvdata(pdev, priv);

	dev_info(dev, "device created");

	/* Allocate both buffers */
	priv->cipher_buf = kzalloc(AES_BLOCK_SIZE, GFP_KERNEL);
	BUG_ON(!priv->cipher_buf);
	priv->plain_buf = kzalloc(AES_BLOCK_SIZE, GFP_KERNEL);
	BUG_ON(!priv->plain_buf);

	priv->regs = ioremap(AES_BASE, AES_SIZE);

	/* Map ciphertext buffer */
	priv->cipher_dma = dma_map_single(&pdev->dev,
				priv->cipher_buf, AES_BLOCK_SIZE, DMA_TO_DEVICE);
	BUG_ON(dma_mapping_error(&pdev->dev, priv->cipher_dma));

	iowrite32(priv->cipher_dma, &priv->regs->dma_read);

	/* Map plaintext buffer */
	priv->plain_dma = dma_map_single(&pdev->dev,
				priv->plain_buf, AES_BLOCK_SIZE, DMA_FROM_DEVICE);
	BUG_ON(dma_mapping_error(&pdev->dev, priv->plain_dma));

	iowrite32(priv->plain_dma, &priv->regs->dma_write);

	err = sysfs_create_group(&dev->kobj, &aes_attr_group);
	BUG_ON(err);

	dev_info(dev, "files created");

	return 0;
}

static int aes_remove(struct platform_device *pdev)
{
	struct aes_priv *priv;

	priv = platform_get_drvdata(pdev);

	device_unregister(priv->dev);

	iounmap(priv->regs);

	dma_unmap_single(&pdev->dev, priv->cipher_dma, AES_BLOCK_SIZE, DMA_TO_DEVICE);
	dma_unmap_single(&pdev->dev, priv->plain_dma, AES_BLOCK_SIZE, DMA_FROM_DEVICE);

	kfree(priv->cipher_buf);
	kfree(priv->plain_buf);

	dev_info(&pdev->dev, "device removed");

	return 0;
}

static const struct of_device_id aes_id_table[] = {
	{.compatible = "stcmtk,aes"},
	{}
};
MODULE_DEVICE_TABLE(of, aes_id_table);

static struct platform_driver aes_drv = {
	.probe  = aes_probe,
	.remove = aes_remove,
	.driver = {
		.name = "aes",
		.of_match_table = aes_id_table,
	}
};

static int aes_init(void)
{
	int err;

	encryptor_class = class_create(THIS_MODULE, "encryptor");
	BUG_ON(IS_ERR(encryptor_class));
	pr_info("class created\n");

	err = platform_driver_register(&aes_drv);
	BUG_ON(err);

	return 0;
}
module_init(aes_init);

static void aes_exit(void)
{
	platform_driver_unregister(&aes_drv);

	class_destroy(encryptor_class);
	pr_info("class destroyed\n");
}
module_exit(aes_exit);

MODULE_AUTHOR("Ivan Oleynikov");
MODULE_LICENSE("GPL");


aes-crypto-api-iface.c
#include <linux/module.h>
#include <linux/kernel.h>
#include <linux/platform_device.h>
#include <linux/of.h>
#include <linux/printk.h>
#include <linux/device.h>
#include <linux/of_irq.h>
#include <linux/interrupt.h>
#include <crypto/algapi.h>
#include <crypto/aes.h>
#include <crypto/padlock.h>
#include <linux/scatterlist.h>

#include <linux/highmem.h>
#include <linux/dma-mapping.h>
#include <linux/slab.h>

#include <linux/delay.h>
#include <linux/jiffies.h>

#include <linux/kdev_t.h>
#include "netdma.h"

#define FPGASLAVES (0xC0000000)

#define DMA_BASE (FPGASLAVES)
#define DMA_SIZE (0x18)

#define AES_BASE  (FPGASLAVES + 0x2000)
#define AES_SIZE  (10 * 4)

#define DECRYPT_OFFSET (0x0000)
#define ENCRYPT_OFFSET (0x4000)

#define AES_BLOCK_SIZE 16
#define FPGA_AUXDATA    8
#define AES_KEY_SIZE   (32 * 4 / 8)

struct aes_regs {
	u32 main_ctrl;
	u32 key[4];
	u32 iv[4];
	u32 block_counter;
} __attribute__ ((__packed__));

#define NETDMA_CSR_SIZE 32

struct netdma_regs {
	u32 control;
	u32 status;
	u32 tx_status;
	u32 rx_report;

	u32 src_desc;
	u32 dst_desc;
} __attribute__ ((packed, aligned(NETDMA_CSR_SIZE)));

struct sg_meta_info {
	void *from;
	struct page *to_page;
	ssize_t to_offset;
	ssize_t len;
	bool last;
};

/* This structure holds everypthing that is necessary to access an instance of
 * hardware crypto accelerator.  It it instantiated twise for encryption and
 * decryption. */
struct aes_priv_hwinfo {
	struct aes_regs __iomem *aes_regs;
	struct netdma_regs __iomem *dma_regs;

	int irq;
};

struct aes_priv {
	uint32_t old_seq;
	struct device *dev;

	/* These fields are used by both encryption and decryption functions.
	 * Looks like kernel crypto subsystem doesn't call them simultaneously
	 * so sharing fields is ok. */
	wait_queue_head_t irq_queue;
	int irq_done;

	struct aes_priv_hwinfo enc, dec;

	struct sg_table dst_table;
	struct page *dst_page;
	dma_addr_t dst_dma;
	void *dst;

	struct sg_table dst_orig_table;

	struct sg_table src_table;
	struct page *src_page;
	dma_addr_t src_dma;
	void *src;

	struct sg_meta_info *meta;
};

struct aes_priv *priv;

static int write_fpga_desc(struct aes_priv *priv, struct netdma_regs __iomem *regs,
		u32 dma_address, u16 length, u8 irq_is_en, u8 is_dst)
{
	u32 control_field;

	control_field = (length << DESC_BYTECOUNT_OFFSET) |
	    (!irq_is_en << DESC_DISABLE_IRQ_OFFSET);

	if (ioread32(&regs->status) & STAT_TX_DESC_BUFFER_FULL) {
		pr_err("%s descriptor buffer full bit is set. Address = 0x%x\n",
		       is_dst ? "rx" : "tx", dma_address);
		return -ENOMEM;
	}

	if (is_dst) {
		iowrite32(dma_address, &regs->dst_desc);
		iowrite32(control_field, &regs->dst_desc);
	} else {
		iowrite32(dma_address, &regs->src_desc);
		iowrite32(control_field, &regs->src_desc);
	}

	wmb();
	return 0;
}

static int fpga_set_key(struct crypto_tfm *tfm, const u8 *in_key,
			unsigned int key_len)
{
	int i;
	const uint32_t *w_buf;

	if (key_len != AES_KEY_SIZE) {
		printk(KERN_ERR "Provided key of length %u when %u expected\n",
		       (unsigned int)key_len, (unsigned int)AES_KEY_SIZE);
		return -EINVAL;
	}

	w_buf = (const uint32_t *)in_key;

	for (i = 3; i >= 0; i--)
		iowrite32(w_buf[i], priv->dec.aes_regs->key + i);

	for (i = 3; i >= 0; i--)
		iowrite32(w_buf[i], priv->enc.aes_regs->key + i);

	return 0;
}

static int fpga_write_iv(const u8 *iv, struct aes_priv_hwinfo *hw)
{
	int i;
	const uint32_t *w_buf;

	w_buf = (const uint32_t *)iv;

	for (i = 3; i >= 0; i--)
		iowrite32(w_buf[i], hw->aes_regs->iv + i);

	return 0;
}

static int fpga_aes_init(struct crypto_tfm *tfm)
{
	return 0;
}

static void fpga_aes_exit(struct crypto_tfm *tfm)
{
}

static void sg_copy_back(struct sg_meta_info *meta)
{
	int i;

	for (i = 0; !meta[i].last; ++i) {
		void *to;

		to = kmap_atomic(meta[i].to_page);

		memcpy(to + meta[i].to_offset, meta[i].from, meta[i].len);

		kunmap_atomic(to);
	}
}

static void set_meta(struct sg_meta_info *meta, struct page *to_page,
		ssize_t to_offset, void *from, ssize_t len)
{
	meta->to_page = to_page;
	meta->to_offset = to_offset;
	meta->from = from;
	meta->len = len;
}

static void sg_split_to_aligned(void *buff, struct page *page,
				struct scatterlist *from, struct scatterlist *to,
				struct sg_meta_info *meta)
{
	struct scatterlist *sg;
	struct scatterlist *old_to;
	ssize_t buff_offset;
	int meta_idx;

	old_to = NULL;

	buff_offset = 0;
	meta_idx = 0;

	for (sg = from; sg; sg = sg_next(sg)) {

		if (!sg->length)
			continue;

		if (sg->length % AES_BLOCK_SIZE) {
			unsigned int first_len, second_len, third_len;
			struct scatterlist *sgn;
			void *sgn_page_ptr, *sg_page_ptr;

			sgn = sg_next(sg);
			sgn_page_ptr = kmap_atomic(sg_page(sgn));
			BUG_ON(!sgn_page_ptr);

			sg_page_ptr = kmap_atomic(sg_page(sg));
			BUG_ON(!sg_page_ptr);

			first_len = sg->length & ~0xf;
			second_len = sg->length & 0xf;
			third_len = AES_BLOCK_SIZE - second_len;

			if (first_len) {
				sg_set_page(to, sg_page(sg), first_len, sg->offset);
				old_to = to;
				to = sg_next(to);
			}

			if (meta) {
				set_meta(&meta[meta_idx++], sg_page(sg), sg->offset + first_len, buff + buff_offset, second_len);
				set_meta(&meta[meta_idx++], sg_page(sgn), sgn->offset, buff + buff_offset + second_len, third_len);
			} else {
				memcpy(buff + buff_offset, sg_page_ptr + sg->offset + first_len, second_len);
				memcpy(buff + buff_offset + second_len, sgn_page_ptr + sgn->offset, third_len);
			}

			sg_set_page(to, page, AES_BLOCK_SIZE, buff_offset);
			old_to = to;
			to = sg_next(to);

			buff_offset += AES_BLOCK_SIZE;

			sgn->offset += third_len;
			sgn->length -= third_len;

			kunmap_atomic(sg_page_ptr);
			kunmap_atomic(sgn_page_ptr);
		} else {
			sg_set_page(to, sg_page(sg), sg->length, sg->offset);
			old_to = to;
			to = sg_next(to);
		}
	}

	if (meta)
		meta[meta_idx].last = 1;
	sg_mark_end(old_to);
}

static void sg_map_all(struct device *dev, struct scatterlist *sg,
		       enum dma_data_direction dir)
{
	struct scatterlist *i;

	for (i = sg; i; i = sg_next(i)) {
		i->dma_address = dma_map_page(dev, sg_page(i), i->offset, i->length, dir);
		BUG_ON(dma_mapping_error(dev, i->dma_address));
	}
}

static void sg_unmap_all(struct device *dev, struct scatterlist *sg,
			 enum dma_data_direction dir)
{
	struct scatterlist *i;

	for (i = sg; i; i = sg_next(i))
		dma_unmap_page(dev, i->dma_address, i->length, dir);
}

static void sg_feed_all(struct aes_priv *priv, struct scatterlist *sg,
		struct aes_priv_hwinfo *hw, bool is_dst)
{
	struct scatterlist *i;
	int err;

	for (i = sg; i; i = sg_next(i)) {
		bool irq_en;

		irq_en = sg_is_last(i) && is_dst;

		err = write_fpga_desc(priv, hw->dma_regs, i->dma_address, i->length, irq_en, is_dst);
		if (err)
			pr_err("write_dst_desc failed: %d\n", err);
	}
}

#define SG_MAX_SIZE 20

static int fpga_crypt(struct blkcipher_desc *desc, struct scatterlist *dst,
			struct scatterlist *src, unsigned int nbytes,
			struct aes_priv_hwinfo *hw)
{
	int err;
	struct scatterlist *i;
	struct scatterlist *src_sg;
	struct scatterlist *dst_sg;
	struct scatterlist *dst_orig_sg;
	struct scatterlist *old_dst_orig_sg;

	BUG_ON(nbytes > PAGE_SIZE);
	BUG_ON(nbytes % AES_BLOCK_SIZE != 0);

	src_sg = priv->src_table.sgl;
	dst_sg = priv->dst_table.sgl;
	dst_orig_sg = priv->dst_orig_table.sgl;

	fpga_write_iv(desc->info, hw);

	priv->irq_done = 0;

	sg_init_table(src_sg, SG_MAX_SIZE);
	sg_init_table(dst_sg, SG_MAX_SIZE);
	sg_init_table(dst_orig_sg, SG_MAX_SIZE);

	old_dst_orig_sg = dst;

	for (i = dst; i; i = sg_next(i)) {
		sg_set_page(dst_orig_sg, sg_page(i), i->length, i->offset);
		old_dst_orig_sg = dst_orig_sg;
		dst_orig_sg = sg_next(dst_orig_sg);
	}

	sg_mark_end(old_dst_orig_sg);

	dst_orig_sg = priv->dst_orig_table.sgl;

	/* Align scatterlists provided to us */
	sg_split_to_aligned(priv->src, priv->src_page, src, src_sg, NULL);
	sg_split_to_aligned(priv->dst, priv->dst_page, dst_orig_sg, dst_sg, priv->meta);

	/* Map memory chunk for passing to DMA controller */
	sg_map_all(priv->dev, src_sg, DMA_TO_DEVICE);
	sg_map_all(priv->dev, dst_sg, DMA_FROM_DEVICE);

	/* Start decryption by writing descriptors */
	sg_feed_all(priv, dst_sg, hw, 1);
	sg_feed_all(priv, src_sg, hw, 0);

	/* Wait for completion interrupt */
	err = wait_event_interruptible(priv->irq_queue, priv->irq_done == 1);
	if (err) {
		printk(KERN_ERR "wait_event_interruptible failed.\n");
		return err;
	}

	/* Unmap chunks back */
	sg_unmap_all(priv->dev, src_sg, DMA_TO_DEVICE);
	sg_unmap_all(priv->dev, dst_sg, DMA_FROM_DEVICE);

	sg_copy_back(priv->meta);

	return err;
}


static int fpga_encrypt(struct blkcipher_desc *desc,
			struct scatterlist *dst,
			struct scatterlist *src, unsigned int nbytes)
{
	return fpga_crypt(desc, dst, src, nbytes, &priv->enc);
}


static int fpga_decrypt(struct blkcipher_desc *desc,
			struct scatterlist *dst,
			struct scatterlist *src, unsigned int nbytes)
{
	return fpga_crypt(desc, dst, src, nbytes, &priv->dec);
}


struct crypto_alg fpga_alg = {
	.cra_name = "cbc(aes)",
	.cra_driver_name = "cbc(aes-fpga)",
	.cra_priority = 1000,
	.cra_flags = CRYPTO_ALG_TYPE_BLKCIPHER,
	.cra_blocksize = AES_BLOCK_SIZE,
	.cra_ctxsize = sizeof(struct aes_priv),
	.cra_type = &crypto_blkcipher_type,
	.cra_alignmask = 15,
	.cra_module = THIS_MODULE,
	.cra_init = fpga_aes_init,
	.cra_exit = fpga_aes_exit,
	.cra_blkcipher = {
		.min_keysize = AES_KEY_SIZE,
		.max_keysize = AES_KEY_SIZE,
		.ivsize = AES_BLOCK_SIZE,
		.setkey = fpga_set_key,
		.encrypt = fpga_encrypt,
		.decrypt = fpga_decrypt,
	}
};

static irqreturn_t fpga_isr(int irq, void *dev_id)
{
	struct aes_priv *priv = dev_id;

	priv->irq_done = 1;
	wake_up_interruptible(&priv->irq_queue);

	return IRQ_HANDLED;
}

static int aes_probe(struct platform_device *pdev)
{
	int err;

	dev_info(&pdev->dev, "probing");

	priv = devm_kzalloc(&pdev->dev, sizeof(*priv), GFP_KERNEL);
	BUG_ON(!priv);

	priv->meta = devm_kzalloc(&pdev->dev, sizeof(struct sg_meta_info) * SG_MAX_SIZE * 2, GFP_KERNEL);
	BUG_ON(!priv->meta);

	priv->dev = &pdev->dev;

	priv->dec.irq = irq_of_parse_and_map(pdev->dev.of_node, 0);
	BUG_ON(!priv->dec.irq);

	priv->enc.irq = irq_of_parse_and_map(pdev->dev.of_node, 1);
	BUG_ON(!priv->enc.irq);

	dev_info(&pdev->dev, "decrypt irq = %d", priv->dec.irq);
	dev_info(&pdev->dev, "encrypt irq = %d", priv->enc.irq);

	priv->dec.aes_regs = ioremap(AES_BASE + DECRYPT_OFFSET, AES_SIZE);
	priv->dec.dma_regs = ioremap(DMA_BASE + DECRYPT_OFFSET, DMA_SIZE);

	priv->enc.aes_regs = ioremap(AES_BASE + ENCRYPT_OFFSET, AES_SIZE);
	priv->enc.dma_regs = ioremap(DMA_BASE + ENCRYPT_OFFSET, DMA_SIZE);

	priv->src = (void *)devm_get_free_pages(priv->dev, GFP_KERNEL, 0);
	priv->dst = (void *)devm_get_free_pages(priv->dev, GFP_KERNEL, 0);

	priv->src_page = virt_to_page(priv->src);
	priv->dst_page = virt_to_page(priv->dst);

	priv->src_dma = dma_map_page(priv->dev, priv->src_page, 0, PAGE_SIZE,
			DMA_TO_DEVICE);
	priv->dst_dma = dma_map_page(priv->dev, priv->dst_page, 0, PAGE_SIZE,
			DMA_FROM_DEVICE);

	err = sg_alloc_table(&priv->src_table, SG_MAX_SIZE, GFP_KERNEL);
	BUG_ON(err);
	err = sg_alloc_table(&priv->dst_table, SG_MAX_SIZE, GFP_KERNEL);
	BUG_ON(err);
	err = sg_alloc_table(&priv->dst_orig_table, SG_MAX_SIZE, GFP_KERNEL);
	BUG_ON(err);

	iowrite32(0, &priv->dec.aes_regs->main_ctrl);
	iowrite32(1, &priv->dec.aes_regs->main_ctrl);
	iowrite32(0, &priv->dec.aes_regs->main_ctrl);

	iowrite32(6, &priv->dec.dma_regs->control);

	iowrite32(0, &priv->enc.aes_regs->main_ctrl);
	iowrite32(1, &priv->enc.aes_regs->main_ctrl);
	iowrite32(0, &priv->enc.aes_regs->main_ctrl);

	iowrite32(6, &priv->enc.dma_regs->control);

	init_waitqueue_head(&priv->irq_queue);

	err = request_irq(priv->dec.irq, fpga_isr, IRQF_SHARED, "fpga-aes-decrypt", priv);
	if (err) {
		dev_err(&pdev->dev, "request_irq for encrypt failed!");
		return -ENOMEM;
	}

	err = request_irq(priv->enc.irq, fpga_isr, IRQF_SHARED, "fpga-aes-encrypt", priv);
	if (err) {
		dev_err(&pdev->dev, "request_irq for encrypt failed!");
		return -ENOMEM;
	}

	err = crypto_register_alg(&fpga_alg);
	BUG_ON(err);

	return 0;
}

static int aes_remove(struct platform_device *pdev)
{
	crypto_unregister_alg(&fpga_alg);

	free_irq(priv->dec.irq, priv);
	free_irq(priv->enc.irq, priv);

	sg_free_table(&priv->src_table);
	sg_free_table(&priv->dst_table);
	sg_free_table(&priv->dst_orig_table);

	dma_unmap_page(priv->dev, priv->src_dma, PAGE_SIZE, DMA_TO_DEVICE);
	dma_unmap_page(priv->dev, priv->dst_dma, PAGE_SIZE, DMA_FROM_DEVICE);

	iounmap(priv->dec.aes_regs);
	iounmap(priv->dec.dma_regs);

	iounmap(priv->enc.aes_regs);
	iounmap(priv->enc.dma_regs);

	dev_info(&pdev->dev, "device removed");

	return 0;
}

static const struct of_device_id aes_id_table[] = {
	{.compatible = "stcmtk,aes"},
	{}
};

MODULE_DEVICE_TABLE(of, aes_id_table);

static struct platform_driver aes_drv = {
	.probe = aes_probe,
	.remove = aes_remove,
	.driver = {
		.name = "aes",
		.of_match_table = aes_id_table,
	}
};

module_platform_driver(aes_drv);

MODULE_AUTHOR("Denis Gabidullin");
MODULE_AUTHOR("Ivan Oleynikov");
MODULE_LICENSE("GPL");
